core/base/perturbation.cpp

```diff
@@ -0,0 +1,91 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/perturbation.hpp>
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+
+
+template <typename ValueType>
+void Perturbation<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    // x = (I + scalar * basis * projector) * b
+    // temp = projector * b                 : projector->apply(b, temp)
+    // x = b                                : x->copy_from(b)
+    // x = 1 * x + scalar * basis * temp    : basis->apply(scalar, temp, 1, x)
+    using vec = gko::matrix::Dense<ValueType>;
+    auto exec = this->get_executor();
+    auto temp = vec::create(
+        exec, gko::dim<2>(this->projector_->get_size()[0], b->get_size()[1]));
```

Comment:
I would advise you to use a cache inside your class. That would eliminate multiple allocations / deallocations in case the same `b` size is used.
An example can be found in [composition.hpp](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/composition.hpp#L152).
Note: I would reccomend to also create this mutable struct, especially including the copy constructor. That way, you don't have to write a copy-constructor and an `operator=` for your `Perturbation` class since it is trivial and will thereby generated for you (by the compiler).
You also need to add a `@note` in the class that `apply` is not thread safe (since it uses an internal cache).